tests/RobotFramework/tests/cumulocity/mapper_recovery/recover_and_publish_software_update_message.robot

```diff
@@ -0,0 +1,41 @@
+*** Settings ***
+Resource    ../../../resources/common.resource
+Library    Cumulocity
+Library    ThinEdgeIO
+
+Test Tags    theme:c8y    theme:mapper recovery 
+Suite Setup    Custom Setup
+Test Teardown    Get Logs
+
+*** Test Cases ***
+Mapper recovers and processes output of ongoing software update request
+    # Mapper receieves a software update request
+    # Delegates operation to agent and gets back `executing` status
+    # And then goes down (here purposefully stopped)
+    # Mean while the agent processes the update message and publishes the software update message
+    # After some time mapper recovers and pushes the result to c8y cloud
+    # Verify that the rolldice package is installed or not
+    ThinEdgeIO.Service Should Be Running    tedge-mapper-c8y    
+    ${OPERATION}=    Install Software        rolldice
+    Should Have MQTT Messages    tedge/commands/res/software/update
+    ThinEdgeIO.Stop Service    tedge-mapper-c8y    
+    Should Have MQTT Messages    tedge/commands/res/software/update
+    ThinEdgeIO.Start Service    tedge-mapper-c8y
```

Comment:
The calls to `Should Have MQTT Messages` need to be a bit more specific, as the two calls can match on the same message published to `tedge/commands/res/software/update` as you are only checking that "a" message exists, but technically it could match either the `executing` update or the `successful` update.

You can add some message level filtering in `Should Have MQTT Messages` to ensure you are actually checking the expected messages.

```
Should Have MQTT Messages tedge/commands/res/software/update    message_contains=executing
ThinEdgeIO.Stop Service tedge-mapper-c8y
Should Have MQTT Messages tedge/commands/res/software/update    message_contains=successful
```

Reply:
Though a general comment. What happens if we restart the `tedge-agent` whilst the software is being processed? Will the software management plugin be called again by the `tedge-agent` when it restarts?

Reply:
I did not find `tedge-agent` calling the `sw management plugin` again.